### PR TITLE
if one check is successful, other checks are now cancelled

### LINKED ISSUE #40: Cancel outstanding HTTP requests on DNS success
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/is-online/issues/40)

Adding this issue here so we don't forget about it.

Details in https://github.com/sindresorhus/is-online/pull/39#issuecomment-279813937, it' currently blocked by https://github.com/sindresorhus/got/issues/274.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $40.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($40.00)

### Submitted pull Requests
- [#65 if one check is successful, other checks are now cancelled](https://issuehunt.io/r/sindresorhus/is-online/pull/65)
- [#69 Query leak](https://issuehunt.io/r/sindresorhus/is-online/pull/69)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/is-online/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by adityapatadia**
closes #40

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#40: Cancel outstanding HTTP requests on DNS success](https://issuehunt.io/repos/24239833/issues/40)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

**Comment by adityapatadia** [comment]
Tests are failing because `.then` is used which can't be avoided.

**Comment by sindresorhus** [comment]
> Tests are failing because .then is used which can't be avoided.

It can always be avoided.

File: index.js [hunk 1]
```diff
@@ -1,40 +1,59 @@
...
 	});
 
-	return /Success/.test(body || '') || Promise.reject();
+	const promise = gotPromise.then(res => {
+		return /Success/.test(res.body || '') || Promise.reject();
+	}).catch(error => {
+		if (!(error instanceof got.CancelError)) {
+			throw error;
+		}
+	});
```
Comment (sindresorhus) [inline index.js]:
```js
const promise = (async () => {
	try {
		const {body} = await gotPromise;
		if (!/Success/.test(body || '')) {
			throw new Error();
		}
	} catch (error) {
		if (!(error instanceof got.CancelError)) {
			throw error;
		}
	}
})();
```

File: index.js [hunk 2]
```diff
@@ -1,40 +1,59 @@
...
 	]);
 
-	return pTimeout(promise, options.timeout).catch(() => false);
+	try {
+		await pTimeout(promise, options.timeout);
+		dnsPromise.cancel();
+		httpsPromise.cancel();
+		applePromise.cancel();
+		return true;
+	} catch (error) {
```
Comment (sindresorhus) [inline index.js]:
```suggestion
	} catch (_) {
```

**Comment by sindresorhus** [comment]
This needs some tests.

**Comment by sindresorhus** [comment]
@adityapatadia Bump

**Pull request closed without merging**